sample-apps/discounts-tutorial/web/frontend/pages/Volume/[id].jsx

```diff
@@ -0,0 +1,305 @@
+          ? DiscountMethod.Automatic
+          : DiscountMethod.Code
+      ),
+      discountCode: useField(discount.code || ""),
+      combinesWith: useField(discount.combinesWith || {}),
+      usageLimit: useField(discount.usageLimit),
+      appliesOncePerCustomer: useField(discount.appliesOncePerCustomer),
+      startsAt: useField(discount.startsAt || null),
+      endsAt: useField(discount.endsAt || null),
+      configuration: {
+        quantity: useField(discount.configuration?.quantity || 1),
+        percentage: useField(discount.configuration?.percentage || 0),
+      },
+    },
+    onSubmit: async (form) => {
+      const newDiscount = {
+        combinesWith: form.combinesWith,
+        startsAt: form.startsAt,
+        endsAt: form.endsAt,
+        metafields: [
+          {
+            id: discount.configurationId, // metafield id is required for update
+            namespace: METAFIELD_NAMESPACE,
+            key: METAFIELD_CONFIGURATION_KEY,
+            type: "json",
+            value: JSON.stringify({
+              quantity: parseInt(form.configuration.quantity),
+              percentage: parseFloat(form.configuration.percentage),
+            }),
+          },
+        ],
+      };
+
+      if (form.discountMethod === DiscountMethod.Code) {
+        newDiscount.usageLimit = parseInt(form.usageLimit);
+        newDiscount.appliesOncePerCustomer = form.appliesOncePerCustomer;
+        newDiscount.code = form.code;
+        newDiscount.title = form.discountCode;
+        newDiscount.code = form.discountCode;
+      } else newDiscount.title = form.discountTitle;
+
+      try {
+        await updateDiscount({ id, discount: newDiscount });
+
+        setSuccessActive(true);
+
+        return { status: "success" };
+      } catch (err) {
+        setErrors(err);
+
+        return { status: "fail" };
+      }
+    },
+  });
+
+  const isUpdateInProgress =
+    isLoading || submitting || updateInProgress || deleteInProgress;
+
+  const isSaveDisabled = isUpdateInProgress || !dirty;
+
+  const handleDeleteDiscount = async () => {
+    if (isUpdateInProgress) return;
+
+    try {
+      await deleteDiscount({ id });
+
+      onBreadcrumbAction(redirect, true);
+    } catch (err) {
+      setErrors(err);
+    }
+  };
+
+  const toggleSuccessActive = () =>
+    setSuccessActive((successActive) => !successActive);
+
+  const toggleDeleteModalActive = () =>
+    setDeleteModalActive((deleteModalActive) => !deleteModalActive);
+
+  const successBanner = successActive ? (
+    <Layout.Section>
+      <Banner
+        title="Discount saved"
+        status="success"
+        onDismiss={toggleSuccessActive}
+      />
+    </Layout.Section>
+  ) : null;
+
+  const errorBanner =
+    errors.length > 0 ? (
+      <Layout.Section>
+        <Banner status="critical">
+          <p>There were some issues with your form submission:</p>
+          <ul>
+            {errors.map(({ message, field }, index) => {
```

Comment:
I think there are cases where this isn't a `UserError` so the `field.join` will fail. For example, [here](https://github.dev/Shopify/function-examples/blob/e7d69934d0dae0ad94f89342b17416137cadd70a/sample-apps/discounts-tutorial/web/frontend/hooks/useUpdateDiscount.js#L67) we reject with an error which would be set [here](https://github.dev/Shopify/function-examples/blob/e7d69934d0dae0ad94f89342b17416137cadd70a/sample-apps/discounts-tutorial/web/frontend/pages/Volume/%5Bid%5D.jsx#L120). We should probably only do `field.join` if `field` exists. Otherwise, it's probably fine to just print the message.